pkg/syncer/specsyncer.go

```diff
@@ -77,58 +76,65 @@ func NewSpecSyncer(from, to *rest.Config, syncedResourceTypes []string, clusterI
+	// TODO: get UID of just-deleted object and pass it as a precondition on this delete.
+	// This would avoid races where an object is deleted and another object with the same name is created immediately after.
+
+	return c.getClient(gvr, namespace).Delete(ctx, name, metav1.DeleteOptions{})
 }
 
 // TODO:
 // This function is there as a quick and dirty implementation of namespace creation.
 // In fact We should also be getting notifications about namespaces created upstream and be creating downstream equivalents.
-func (c *Controller) ensureNamespaceExists(namespace string) error {
+func (c *Controller) ensureNamespaceExists(ctx context.Context, namespace string) error {
 	namespaces := c.toClient.Resource(schema.GroupVersionResource{
 		Group:    "",
 		Version:  "v1",
 		Resource: "namespaces",
 	})
+
+	// See if we need to create the target namespace
+	_, err := namespaces.Get(ctx, namespace, metav1.GetOptions{})
+	if err == nil {
+		// Already exists - all good
+		return nil
+	}
+
+	if !k8serrors.IsNotFound(err) {
+		// Got some error besides not-found - stop & return.
+		// TODO bubble this up as a condition somewhere.
+		klog.Errorf("Error checking if namespace %q exists: %v", namespace, err)
+		return err
+	}
+
+	// Not found - try to create
 	newNamespace := &unstructured.Unstructured{}
 	newNamespace.SetAPIVersion("v1")
 	newNamespace.SetKind("Namespace")
 	newNamespace.SetName(namespace)
 	if _, err := namespaces.Create(context.TODO(), newNamespace, metav1.CreateOptions{}); err != nil {
+		// An already exists error is ok - it means something else beat us to creating the namespace.
 		if !k8serrors.IsAlreadyExists(err) {
-			klog.Infof("Error while creating namespace %s: %v", namespace, err)
+			// Any other error is not good, though.
+			// TODO bubble this up as a condition somewhere.
+			klog.Errorf("Error while creating namespace %q: %v", namespace, err)
 			return err
 		}
 	}
-	return nil
-}
-
-func deleteFromDownstream(c *Controller, ctx context.Context, gvr schema.GroupVersionResource, namespace, name string) error {
-	// TODO: get UID of just-deleted object and pass it as a precondition on this delete.
-	// This would avoid races where an object is deleted and another object with the same name is created immediately after.
 
-	return c.getClient(gvr, namespace).Delete(ctx, name, metav1.DeleteOptions{})
+	return nil
 }
 
 func upsertIntoDownstream(c *Controller, ctx context.Context, gvr schema.GroupVersionResource, namespace string, unstrob *unstructured.Unstructured) error {
-	if err := c.ensureNamespaceExists(namespace); err != nil {
-		klog.Error(err)
+	if err := c.ensureNamespaceExists(ctx, namespace); err != nil {
 		return err
 	}
 
 	client := c.getClient(gvr, namespace)
 
-	unstrob = unstrob.DeepCopy()
-
 	// Attempt to create the object; if the object already exists, update it.
 	unstrob.SetUID("")
 	unstrob.SetResourceVersion("")
```

Comment:
is this the only change we do to the object when copying from up to downstream?

It's not enough. We have to clean much more. Finalizers come to mind, but also annotations that control aspects. IMO we should strip all annotations, all owner refs too. Maybe more. cc @marun

Reply:
Also note line 160. We cannot do that. An OpenShift will use annotation, maybe other actors as well. We cannot just wipe all changes. That will also hotloop.